easycv/core/optimizer/adam.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (c) Alibaba, Inc. and its affiliates.
+import math
+from typing import List
+
+import torch
+from mmcv.runner.optimizer.builder import OPTIMIZERS
+from torch import Tensor
+from torch.optim import *  # noqa: F401,F403
+from torch.optim.optimizer import Optimizer
+
+
+def adamw(params: List[Tensor], grads: List[Tensor], exp_avgs: List[Tensor],
+          exp_avg_sqs: List[Tensor], max_exp_avg_sqs: List[Tensor],
+          state_steps: List[int], amsgrad: bool, beta1: float, beta2: float,
+          lr: float, weight_decay: float, eps: float):
+    r"""Functional API that performs AdamW algorithm computation.
+    See :class:`~torch.optim.AdamW` for details.
+    """
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        step = state_steps[i]
+
+        # Perform stepweight decay
+        param.mul_(1 - lr * weight_decay)
+
+        bias_correction1 = 1 - beta1**step
+        bias_correction2 = 1 - beta2**step
+
+        # Decay the first and second moment running average coefficient
+        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+        exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+        if amsgrad:
+            # Maintains the maximum of all 2nd moment running avg. till now
+            torch.maximum(
+                max_exp_avg_sqs[i], exp_avg_sq, out=max_exp_avg_sqs[i])
+            # Use the max. for normalizing running avg. of gradient
+            denom = (max_exp_avg_sqs[i].sqrt() /
+                     math.sqrt(bias_correction2)).add_(eps)
+        else:
+            denom = (exp_avg_sq.sqrt() / math.sqrt(bias_correction2)).add_(eps)
+
+        step_size = lr / bias_correction1
+
+        param.addcdiv_(exp_avg, denom, value=-step_size)
+
+
+@OPTIMIZERS.register_module()
+class _AdamW(Optimizer):
```

Comment:
rename `_AdamW` to `AdamW`, and rewritre this api to inherit torch optimizer

pop original registry if name conflict

Reply:
already modified